text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,460 @@
+when they are added.
+
+These two messages should be distinct; the former says "this enum/struct has
+stuff that you shouldn't see," while the latter says "this enum/struct is
+incomplete and may be extended in the future."
+
+# How We Teach This
+
+Changes to rustdoc should make it easier for users to understand the concept of
+non-exhaustive enums and structs in the wild.
+
+In the chapter on enums, a section should be added specifically for
+non-exhaustive enums. Because error types are common in almost all crates, this
+case is important enough to be taught when a user learns Rust for the first
+time.
+
+Additionally, non-exhaustive structs should be documented in an early chapter on
+structs. Public fields should be preferred over getter/setter methods in Rust,
+although users should be aware that adding extra fields is a potentially
+breaking change.
+
+# Drawbacks
+
+* The `#[doc(hidden)]` hack in practice is usually good enough.
+* An attribute may be more confusing than a dedicated syntax.
+* `non_exhaustive` may not be the clearest name.
+
+# Alternatives
+
+* Provide a dedicated syntax instead of an attribute. This would likely be done
+  by adding a `...` variant or field, as proposed by the original
+  [extensible enums RFC][RFC 757].
+* Allow creating private enum variants, giving a less-hacky way to create a
+  hidden variant.
+* Document the `#[doc(hidden)]` hack and make it more well-known.
+
+# Unresolved questions
+
+Should there be a way to warn downstream crate users when their match is
+non-exuhaustive? What if a user wants to add a warning to their matches using
+non-exhaustive enums, to indicate that more code should be added to handle a new
+variant?
+
+## Extending to traits
+
+Tangentially, it also makes sense to have non-exhaustive traits as well, even
+though they'd be non-exhaustive in a different way. Take this example from
+[`byteorder`]:
+
+```rust
+pub trait ByteOrder: Clone + Copy + Debug + Default + Eq + Hash + Ord + PartialEq + PartialOrd {
+   // ...
+}
+```
+
+The `ByteOrder` trait requires these traits so that a user can simply write a
+bound of `T: ByteOrder` without having to add other useful traits, like `Hash`
+or `Eq`.
+
+This trait is useful, but the crate has no intention of letting other users
+implement this trait themselves, because then adding an additional trait
+dependency for `ByteOrder` could be a breaking change.
+
+The way that this crate solves this problem is by adding a hidden trait
+dependency:
+
+```rust
+mod private {
+    pub trait Sealed {}
+    impl Sealed for super::LittleEndian {}
+    impl Sealed for super::BigEndian {}
+}
+
+pub trait ByteOrder: /* ... */ + private::Sealed {
+    // ...
+}
+```
+
+This way, although downstream crates can use this trait, they cannot actually
+implement things for this trait.
+
+This pattern could again be solved by using `#[non_exhaustive]`:
+
+```rust
+#[non_exhaustive]
```

Comment:
I think`no_extern_impl` is a better name, because exhaustivenes feels weird for traits.